README.md
# sv-rollout

`sv-rollout` is used to restart multiple runit services concurrently, a
configurable ratio of them to time out, with configurable concurrency, and
optionally "canaries", which are restarted before the rest of the services.

```
Usage of sv-rollout:
  -canary-ratio=0.001: canary nodes are restarted first. If they fail, the deploy is failed. Rounded up to the nearest node, unless set to zero
  -canary-timeout-tolerance=0: ratio of canary nodes that are permitted to time out without causing the deploy to fail
  -chunk-ratio=0.2: after canary nodes, ratio of remaining nodes permitted to restart concurrently
  -pattern="": (required) glob pattern to match /etc/service entries (e.g. "borg-shopify-*")
  -timeout=90: number of seconds to wait for a service to restart before considering it timed out and moving on
  -timeout-tolerance=0: ratio of total nodes whose restarts may time out and still consider the deploy a success
Examples:
  # Restart one service first. Restart everything else once it succeeds. No timeouts allowed, wait up to 5 minutes for restarts.
  sv-rollout -canary-ratio 0.0001 -chunk-ratio 1 -timeout 300 -pattern 'borg-*'
  # Restart 10% of services first, allowing up to 50% of those to time out. Then, restart all other services, 30% at a time, allowing up to 70% to time out.
  sv-rollout -canary-ratio 0.1 -chunk-ratio 0.3 -canary-timeout-tolerance 0.5 -timeout-tolerance 0.7 -timeout 300 -pattern 'borg-*'
```

## Examples

#### Job servers

Here, we want to restart a small handful first to make sure the new container is
actually good, after which we pretty much want to restart everything ASAP.

We can use canaries for this, but we need to be mindful of the fact that jobs
can take a while to restart. So we pick a small handful of services to restart
first, allowing most of them to time out (but not fail outright). After that, we
just restart everything, allowing most of them to time out (but again, not fail
outright).

```
sv-rollout \
  -canary-ratio 0.1 \             # restart 10% first
  -canary-timeout-tolerance 0.7 \ # allow up to 70% of those to time out
  -chunk-ratio 1 \                # after canaries, restart 100% concurrently
  -timeout-tolerance 0.8 \        # allow up to 80% of remaining to time out
  -timeout 300 \                  # wait up to 300s for each service before considering it "timed out"
  -pattern 'borg-shopify-jobs-*'  # any services matching /etc/service/<pattern>
```

#### App servers

Here, we have five containers, and want to restart them one at a time, accepting
no timeouts. Much simpler!

```
sv-rollout \
  -canary-ratio 0 \                 # no canaries.
  -chunk-ratio 0.2 \                # restart 20% of servers concurrently (ie.  no concurrency)
  -timeout-tolerance 0 \            # even a single timeout will abort the deploy
  -timeout 300 \                    # wait up to 300s for each service before considering it "timed out"
  -pattern 'borg-shopify-unicorn-*' # any services matching /etc/service/<pattern>
```



CHANGELOG.md
# Unreleased

# 1.2.3

* Fix race condition in preempt mode ( https://github.com/Shopify/sv-rollout/issues/9 ) causing panics

# 1.2.2

* Move from Godeps to gb 
* Move source to `src/cmd` as part of gb migration
* Vendor dependencies as git submodules and use Go1.5+ native vendoring support
* Update dependencies
* Add helper scripts for building,running etc in `scripts`
* Remove OSX support

# 1.2.1

* Add support for lockfile in /var/lock/dont-sv-rollout to prohibit sv-rollout from restarting services

# 1.2.0

* Add StatsD instrumentation for restart durations.

# 1.1.1

* Bump results channel size to possibly avoid deadlocks.
* Print to STDERR instead of STDOUT on timeout or failure.

# 1.1.0

* Randomize order of service restarts.
* [Timeout Preemption](https://github.com/Shopify/sv-rollout/pull/6)

# 1.0.3

* Write failures and timeouts to STDERR.
* Add `-oncomplete` to specify a command to run when done.

# 1.0.2

* Write log output to STDOUT instead of STDERR.

# 1.0.1

* Removed .changes generation
* Fixed path to `sv`

# 1.0.0

* Initial release


deployify_test.go
package main

import (
	"path/filepath"
	"testing"

	. "github.com/smartystreets/goconvey/convey"
)

func TestDeployify(t *testing.T) {

	Convey("Enumerating services", t, func() {
		defer func() {
			globServices = filepath.Glob
		}()
		globServices = func(a string) ([]string, error) {
			return []string{"/etc/sv/borg-shopify-test-1", "/etc/sv/borg-shopify-test-2"}, nil
		}

		Convey("Should generate corrcect service names", func() {
			svcs, err := GetServices("borg-shopify-*")
			So(err, ShouldBeNil)
			So(svcs, ShouldResemble, []string{"borg-shopify-test-1", "borg-shopify-test-2"})
		})
	})
}


deployment_test.go
package main

import (
	"sync/atomic"
	"testing"
	"time"

	. "github.com/smartystreets/goconvey/convey"
)

var (
	concurrency int32
)

var (
	quantum = 25 * time.Millisecond
)

func restartWithTiming(svr *SvRestarter, d time.Duration) error {
	atomic.AddInt32(&concurrency, 1)
	time.Sleep(2 * quantum)
	atomic.AddInt32(&concurrency, -1)

	return nil
}

func alwaysPass(svr *SvRestarter, d time.Duration) error {
	return nil
}

func alwaysTimeout(svr *SvRestarter, d time.Duration) error {
	return ErrRestartTimeout{Service: svr.Service}
}

func alwaysFail(svr *SvRestarter, d time.Duration) error {
	return ErrRestartFailed{Service: svr.Service, Message: "failed"}
}

func timeoutOneService(svr *SvRestarter, d time.Duration) error {
	if svr.Service == "b" {
		return alwaysTimeout(svr, d)
	}
	return alwaysPass(svr, d)
}

func failOneService(svr *SvRestarter, d time.Duration) error {
	if svr.Service == "b" {
		return alwaysFail(svr, d)
	}
	return alwaysPass(svr, d)
}

func TestDeployment(t *testing.T) {

	config := config{
		CanaryRatio:            0.0001,
		CanaryTimeoutTolerance: 0,
		ChunkRatio:             0.2,
		TimeoutTolerance:       0,
		Timeout:                1,
	}

	Convey("Running a deployment", t, func() {

		Convey("with no canaries and 50% timeouts allowed", func() {
			config.CanaryRatio = 0
			config.TimeoutTolerance = 0.5
			config.ChunkRatio = 0.001
			Convey("succeeds when everything restarts successfully", func() {
				restartSvr = alwaysPass
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				err := depl.Run()
				So(err, ShouldBeNil)
			})
			Convey("fails when everything times out", func() {
				restartSvr = alwaysTimeout
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				err := depl.Run()
				So(err, ShouldEqual, ErrTooManyTimeouts)
			})
			Convey("fails when everything fails", func() {
				restartSvr = alwaysFail
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				err := depl.Run()
				So(err, ShouldEqual, ErrTooManyFailures)
			})
			Convey("fails when only one service fails", func() {
				restartSvr = failOneService
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				err := depl.Run()
				So(err, ShouldEqual, ErrTooManyFailures)
			})
			Convey("succeeds when only one service times out", func() {
				restartSvr = timeoutOneService
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				err := depl.Run()
				So(err, ShouldBeNil)
			})
		})

		Convey("with one canary followed by 100%", func() {
			config.CanaryRatio = 0.001
			config.ChunkRatio = 1
			config.TimeoutTolerance = 0.5
			Convey("Fails when the canary times out", func() {
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				restartSvr = func(svr *SvRestarter, d time.Duration) error {
					if svr.Service == depl.canaryServices[0] {
						return alwaysTimeout(svr, 0)
					}
					return nil
				}
				err := depl.Run()
				So(err, ShouldEqual, ErrTooManyTimeouts)
			})
			Convey("Succeeds when one non-canary service times out", func() {
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				restartSvr = func(svr *SvRestarter, d time.Duration) error {
					if svr.Service == depl.postCanaryServices[0] {
						return alwaysTimeout(svr, 0)
					}
					return nil
				}
				err := depl.Run()
				So(err, ShouldBeNil)
			})
			Convey("Fails when the canary fails", func() {
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				restartSvr = func(svr *SvRestarter, d time.Duration) error {
					if svr.Service == depl.canaryServices[0] {
						return alwaysFail(svr, 0)
					}
					return nil
				}
				err := depl.Run()
				So(err, ShouldEqual, ErrTooManyFailures)
			})
			Convey("Succeeds when one non-canary service fails", func() {
				depl := NewDeployment([]string{"a", "b", "c"}, config)
				restartSvr = func(svr *SvRestarter, d time.Duration) error {
					if svr.Service == depl.postCanaryServices[0] {
						return alwaysFail(svr, 0)
					}
					return nil
				}
				err := depl.Run()
				So(err, ShouldEqual, ErrTooManyFailures)
			})
		})

	})

	Convey("Running a deployment with canary-ratio 0.001 and chunk-ratio 0.25", t, func() {
		config.CanaryRatio = 0.001
		config.ChunkRatio = 0.25
		Convey("on 8 nodes", func() {
			depl := NewDeployment([]string{"a", "b", "c", "d", "e", "f", "g", "h"}, config)
			restartSvr = restartWithTiming

			Convey("should restart the canary alone, then two nodes concurrently", func() {
				ch := make(chan error)
				go func() {
					ch <- depl.Run()
				}()
				time.Sleep(quantum) // put us out of phase with the sleeps in the restart code
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 1)
				time.Sleep(2 * quantum)
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 2)
				time.Sleep(2 * quantum)
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 2)
				time.Sleep(2 * quantum)
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 2)
				time.Sleep(2 * quantum)
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 1)
				time.Sleep(2 * quantum)
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 0)

				var err error
				select {
				case err = <-ch:
					So(err, ShouldBeNil)
				default:
					// should have been done
					So(nil, ShouldNotBeNil)
				}
			})
		})
	})

	Convey("Running a deployment with canary-ratio 0 and chunk-ratio 0.5", t, func() {
		config.CanaryRatio = 0
		config.ChunkRatio = 0.5
		Convey("on 3 nodes", func() {
			depl := NewDeployment([]string{"a", "b", "c"}, config)
			restartSvr = restartWithTiming

			Convey("should restart two nodes, then the last one", func() {
				ch := make(chan error)
				go func() {
					ch <- depl.Run()
				}()
				time.Sleep(quantum) // put us out of phase with the sleeps in the restart code
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 2)
				time.Sleep(2 * quantum)
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 1)
				time.Sleep(2 * quantum)
				So(atomic.LoadInt32(&concurrency), ShouldEqual, 0)

				var err error
				select {
				case err = <-ch:
					So(err, ShouldBeNil)
				default:
					// should have been done
					So(nil, ShouldNotBeNil)
				}
			})
		})

	})

}


sv-rollout_test.go
package main

import (
	"path/filepath"
	"testing"

	. "github.com/smartystreets/goconvey/convey"
)

func TestSvRollout(t *testing.T) {

	Convey("Shuffling services", t, func() {
		defer func() {
			globServices = filepath.Glob
		}()
		unexpected := []string{"a", "b", "c", "d", "e", "f", "g", "h"}
		globServices = func(a string) ([]string, error) {
			var rets []string
			for _, str := range unexpected {
				rets = append(rets, "/etc/service/"+(str))
			}
			return rets, nil
		}
		svcs, err := getServices("borg-shopify-*")

		Convey("Services should be shuffled", func() {
			So(err, ShouldBeNil)
			So(len(svcs), ShouldEqual, len(unexpected))
			for _, str := range unexpected {
				So(svcs, ShouldContain, str)
			}
			// Tiny possibility of random failure.
			So(svcs, ShouldNotResemble, unexpected)
		})
	})

	Convey("Enumerating services", t, func() {
		defer func() {
			globServices = filepath.Glob
		}()
		globServices = func(a string) ([]string, error) {
			return []string{"/etc/service/borg-shopify-test-1", "/etc/service/borg-shopify-test-2"}, nil
		}
		svcs, err := getServices("borg-shopify-*")

		Convey("Should generate corrcect service names", func() {
			So(err, ShouldBeNil)
			So(len(svcs), ShouldEqual, 2)
			So(svcs, ShouldContain, "borg-shopify-test-1")
			So(svcs, ShouldContain, "borg-shopify-test-2")
		})
	})

	Convey("Choosing canaries", t, func() {
		Convey("should not choose any if ratio = 0", func() {
			c, nc := chooseCanaries([]string{"a", "b", "c", "d"}, 0)
			So(len(c), ShouldEqual, 0)
			So(len(nc), ShouldEqual, 4)
		})
		Convey("should choose one if ratio is small", func() {
			c, nc := chooseCanaries([]string{"a", "b", "c", "d", "e", "f", "g"}, 0.001)
			So(len(c), ShouldEqual, 1)
			So(len(nc), ShouldEqual, 6)
		})
		Convey("Should round up in general", func() {
			c, nc := chooseCanaries([]string{"a", "b", "c", "d", "e"}, 0.5)
			So(len(c), ShouldEqual, 3)
			So(len(nc), ShouldEqual, 2)
		})
	})

	Convey("Deciding on permitted timeouts", t, func() {
		Convey("should not allow any if ratio = 0", func() {
			So(permittedTimeouts([]string{"a", "b", "c"}, 0), ShouldEqual, 0)
		})
		Convey("should allow one if ratio is small", func() {
			So(permittedTimeouts([]string{"a", "b", "c", "d", "e", "f"}, 0.001), ShouldEqual, 1)
		})
		Convey("should allow all if ratio is 1", func() {
			So(permittedTimeouts([]string{"a", "b", "c", "d", "e", "f"}, 1), ShouldEqual, 6)
		})
	})
}


errors.go
package main

import (
	"errors"
	"fmt"
)

// ErrTooManyTimeouts means that the number of services which timed out exceeded
// the threshold, and the deploy should be aborted.
var ErrTooManyTimeouts = errors.New("too many services timed out while restarting")

// ErrTooManyFailures means that a service failed to restart and the deploy
// should be aborted.
var ErrTooManyFailures = errors.New("too many services failed to restart")

// ErrRestartPreempted happens when we terminate the deploy early due to a
// sufficient number of services restarting successfully to consider the deploy
// a success even if every remaining service times out.
type ErrRestartPreempted struct {
	Service string
}

func (e ErrRestartPreempted) Error() string {
	return fmt.Sprintf("service '%s' didn't need to restart in time for the deploy to succeed so we stopped watching it", e.Service)
}

// ErrRestartTimeout indicates that a service restart timed out.
type ErrRestartTimeout struct {
	Service string
}

func (e ErrRestartTimeout) Error() string {
	return fmt.Sprintf("restart timed out for service '%s'", e.Service)
}

// ErrRestartFailed indicates that a service restart failed in a manner other
// than timing out.
type ErrRestartFailed struct {
	Service string
	Message string
}

func (e ErrRestartFailed) Error() string {
	return fmt.Sprintf("restart failed for service '%s': %s", e.Service, e.Message)
}


deployify.go
/*
deployify is a utility to restart multiple runit services concurrently. It
supports canaries and has configurable tolerance for timeouts.
*/
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"path"
	"path/filepath"
)

const (
	svdir = "/etc/sv"
)

type config struct {
	CanaryRatio            float64
	CanaryTimeoutTolerance float64
	ChunkRatio             float64
	TimeoutTolerance       float64
	Timeout                int
}

// Verbose controls whether extra information is printed as choices are made.
// It's set by the `-verbose` CLI flag.
var Verbose bool

func (c config) AssertValid(pattern string) {
	msg := ""
	if pattern == "" {
		msg = "-pattern must be provided"
	}
	if c.ChunkRatio < c.CanaryRatio {
		msg = "-chunk-ratio must be >= -canary-ratio. This is not an inherent limitation, feel free to add code to handle this case."
	}
	if msg != "" {
		fmt.Println(msg)
		flag.Usage()
		os.Exit(1)
	}
}

func init() {
	oldUsage := flag.Usage
	flag.Usage = func() {
		oldUsage()
		fmt.Fprintf(os.Stderr, "Examples:\n")
		fmt.Fprintln(os.Stderr, "  # Restart one service first. Restart everything else once it succeeds. No timeouts allowed, wait up to 5 minutes for restarts.")
		fmt.Fprintln(os.Stderr, "  "+os.Args[0]+" -canary-ratio 0.0001 -chunk-ratio 1 -timeout 300 -pattern 'borg-*'")
		fmt.Fprintf(os.Stderr, "%s", "  # Restart 10% of services first, allowing up to 50% of those to time out. Then, restart all other services, 30% at a time, allowing up to 70% to time out.\n")
		fmt.Fprintln(os.Stderr, "  "+os.Args[0]+" -canary-ratio 0.1 -chunk-ratio 0.3 -canary-timeout-tolerance 0.5 -timeout-tolerance 0.7 -timeout 300 -pattern 'borg-*'")
	}
}

func main() {
	var (
		canaryRatio            = flag.Float64("canary-ratio", 0.001, "canary nodes are restarted first. If they fail, the deploy is failed. Rounded up to the nearest node, unless set to zero")
		canaryTimeoutTolerance = flag.Float64("canary-timeout-tolerance", 0, "ratio of canary nodes that are permitted to time out without causing the deploy to fail")
		chunkRatio             = flag.Float64("chunk-ratio", 0.2, "after canary nodes, ratio of remaining nodes permitted to restart concurrently")
		timeoutTolerance       = flag.Float64("timeout-tolerance", 0, "ratio of total nodes whose restarts may time out and still consider the deploy a success")
		timeout                = flag.Int("timeout", 90, "number of seconds to wait for a service to restart before considering it timed out and moving on")
		pattern                = flag.String("pattern", "", "(required) glob pattern to match /etc/sv entries (e.g. \"borg-shopify-*\")")
		verbose                = flag.Bool("verbose", false, "print more information about what's going on")
	)
	flag.Parse()

	config := config{
		CanaryRatio:            *canaryRatio,
		CanaryTimeoutTolerance: *canaryTimeoutTolerance,
		ChunkRatio:             *chunkRatio,
		TimeoutTolerance:       *timeoutTolerance,
		Timeout:                *timeout,
	}
	config.AssertValid(*pattern)

	Verbose = *verbose

	if Verbose {
		log.Printf("[debug] initializing with pattern=%s, config %#v", *pattern, config)
	}

	os.Exit(run(*pattern, config))
}

func run(servicePattern string, c config) int {
	services, err := getServices(servicePattern)
	if err != nil {
		log.Fatal(err)
	}

	d := NewDeployment(services, c)
	if err := d.Run(); err != nil {
		return 1
	}
	return 0
}

func getServices(pattern string) (services []string, err error) {
	var fullpaths []string
	fullpaths, err = globServices(svdir + "/" + pattern)
	if err != nil {
		return nil, err
	}
	for _, p := range fullpaths {
		services = append(services, path.Base(p))
	}
	return
}

var globServices = filepath.Glob


deployment.go
package main

import (
	"log"
	"math"
	"os"
)

// Deployment orchestrates the concurrent restarting of all the indicated
// services, ultimately returning an error indicating whether it was
// successful.
type Deployment struct {
	numServices int

	canaryServices     []string
	postCanaryServices []string

	canaryTimeoutsPermitted int
	totalTimeoutsPermitted  int

	postCanaryConcurrency int

	timeout int
	index   int

	currentFailuresPermitted int
	currentTimeoutsPermitted int

	successesSoFar int
	timeoutsSoFar  int
	failuresSoFar  int

	svrs []*SvRestarter

	toRestart chan *SvRestarter
	results   chan error
}

// NewDeployment initializes a Deployment object with a list of services
// (entries in /etc/service typically) and a config object.
func NewDeployment(services []string, config config) *Deployment {
	var d Deployment
	d.numServices = len(services)

	d.canaryServices, d.postCanaryServices = chooseCanaries(services, config.CanaryRatio)
	d.canaryTimeoutsPermitted = permittedTimeouts(d.canaryServices, config.CanaryTimeoutTolerance)

	ctp := d.canaryTimeoutsPermitted
	ttp := ctp + permittedTimeouts(d.postCanaryServices, config.TimeoutTolerance)
	d.totalTimeoutsPermitted = ttp

	d.timeout = config.Timeout

	d.postCanaryConcurrency = ceilRatio(d.postCanaryServices, config.ChunkRatio)

	d.toRestart = make(chan *SvRestarter, 8192)
	d.results = make(chan error, 1024)

	if Verbose {
		log.Printf("[debug] chose canaries: %v", d.canaryServices)
		log.Printf("[debug] canaries permitted to time out: %d", d.canaryTimeoutsPermitted)
		log.Printf("[debug] total timeouts permitted: %d", d.totalTimeoutsPermitted)
		log.Printf("[debug] concurrency after canary phase: %d", d.postCanaryConcurrency)
	}

	return &d
}

// Run does all the actual grunt work of concurrently restarting the services.
// It first attempts to restart the services chosen as canaries. If a
// sufficient number of them pass, it will move on to restarting the rest of
// the services with concurrency as indicated by ChunkRatio.
func (d *Deployment) Run() (err error) {
	d.startWorkers(len(d.canaryServices))
	if err = d.restartServices(d.canaryServices, 0, d.canaryTimeoutsPermitted, d.canarySuccessOK); err != nil {
		return
	}
	delta := d.postCanaryConcurrency - len(d.canaryServices)
	d.startWorkers(delta)
	return d.restartServices(d.postCanaryServices, 0, d.totalTimeoutsPermitted, d.allComplete)
}

func (d *Deployment) startWorkers(n int) {
	for i := 0; i < n; i++ {
		go d.startWorker()
	}
}

func (d *Deployment) startWorker() {
	for svr := range d.toRestart {
		d.results <- restartSvr(svr)
	}
}

func (d *Deployment) restartServices(services []string, failuresPermitted, timeoutsPermitted int, done func() bool) (err error) {
	d.currentFailuresPermitted = failuresPermitted
	d.currentTimeoutsPermitted = timeoutsPermitted

	if len(services) == 0 {
		return nil
	}

	// Don't restart services when a lock is present
	if _, err := os.Stat("/var/lock/dont-sv-rollout"); err == nil {
		log.Println("/var/lock/dont-sv-rollout present, not restarting services")
		return nil
	}

	for _, svc := range services {
		d.index++
		svr := NewSvRestarter(svc, d.numServices, d.index, d.timeout)
		d.svrs = append(d.svrs, svr)
		d.toRestart <- svr
	}

	remaining := len(services) // number of services yet to be processed.
	for result := range d.results {
		switch result.(type) {
		case nil:
			d.successesSoFar++
		case ErrRestartFailed:
			if err = d.incrementFailures(); err != nil {
				return
			}
		case ErrRestartTimeout:
			if err = d.incrementTimeouts(); err != nil {
				return
			}
		case ErrRestartPreempted:
			// no need to handle the error here because we pre-verified that it's ok
			// before preempting the svr
			_ = d.incrementTimeouts()
		default:
			panic(result)
		}
		if done() {
			return nil
		}

		if d.currentTimeoutsPermitted >= remaining {
			for _, svr := range d.svrs {
				svr.Preempt()
			}
		}
		remaining--
	}

	panic("unreachable")
}

func (d *Deployment) incrementFailures() error {
	d.failuresSoFar++
	if d.failuresSoFar > d.currentFailuresPermitted {
		return ErrTooManyFailures
	}
	return nil
}

func (d *Deployment) incrementTimeouts() error {
	d.timeoutsSoFar++
	if d.timeoutsSoFar > d.currentTimeoutsPermitted {
		return ErrTooManyTimeouts
	}
	return nil
}

func (d *Deployment) canarySuccessOK() bool {
	canaries := len(d.canaryServices)
	mustPass := canaries - d.canaryTimeoutsPermitted
	return d.successesSoFar >= mustPass
}

func (d *Deployment) allComplete() bool {
	done := d.successesSoFar + d.timeoutsSoFar + d.failuresSoFar
	return done == d.numServices
}

func chooseCanaries(services []string, ratio float64) (canaries []string, nonCanaries []string) {
	nCanary := int(math.Ceil(ratio * float64(len(services))))
	for index, service := range services {
		if index < nCanary {
			canaries = append(canaries, service)
		} else {
			nonCanaries = append(nonCanaries, service)
		}
	}
	return
}

func permittedTimeouts(services []string, tolerance float64) int {
	return ceilRatio(services, tolerance)
}

func ceilRatio(coll []string, ratio float64) int {
	return int(math.Ceil(ratio * float64(len(coll))))
}

// stubbed in tests
var (
	restartSvr = _restartSvr
)

func _restartSvr(svr *SvRestarter) error {
	return svr.Restart()
}


borg/upload_deb.rb
#!/usr/bin/env ruby
# Upload .deb packages to our apt repo

require 'socket'
require 'etc'
require 'tempfile'

class AptUpload
  UPLOAD_USER = 'apt-upload'
  UPLOAD_HOST = 'apt.ec2.shopify.com'

  VALID_ARCHES = %w(amd64)
  VALID_DISTS  = %w(lucid precise trusty utopic)

  def run(*args)
    usage if args.count != 1

    path = args.first
    Dir.entries(path).each do |file|
      upload_changes(File.join(path, file)) if file.end_with?('.changes')
    end
  end

  private

  def usage
    $stderr.puts "Usage: #{$0} <path>"
    exit(1)
  end

  def generate_upload_id
    timestamp = Time.now.utc.strftime("%Y%m%d_%H%M%S")
    username  = 'apt-builder'
    hostname  = Socket.gethostname.split('.', 2).first
    pid       = $$
    random    = rand(16**8)

    id = "%s_%s_%s_%d_%08x" % [timestamp, username, hostname, pid, random]
    raise "Upload ID contains odd characters? #{id.inspect}" if id =~ /[^A-Za-z0-9_-]/
    id
  end

  def upload_changes(file)
    upload_id    = generate_upload_id
    final_path   = "/data/uploads/pending/#{upload_id}"
    temp_path    = final_path + '.tmp'

    ssh_target   = UPLOAD_USER + '@' + UPLOAD_HOST
    rsync_target = ssh_target + ':' + temp_path + '/'

    Dir.mktmpdir(nil, "/tmp") do |tempdir|
      files = process_changes(file, tempdir)
      File.chmod(0755, tempdir)

      puts
      puts "Uploading #{files.count} files to #{temp_path} on #{UPLOAD_HOST} ..."
      sh("rsync", "-dtP", "#{tempdir}/", rsync_target)

      puts
      puts "Queuing for upload ..."
      sh("ssh", ssh_target, "mv '#{temp_path}' '#{final_path}'")

      puts
      puts "Done uploading #{file}."
      sh("sudo", "touch", "#{File.dirname(file)}/upload.stamp")
    end
  end

  def sh(*command)
    begin
      system(*command)
    rescue Exception => e
      $stderr.puts "Error in command: #{command.inspect}"
      raise e
    end

    unless $?.success?
      raise "Command failed with status #{$?.inspect}: #{command.inspect}"
    end
  end

  def process_changes(in_file, tempdir)
    base_path = File.dirname(in_file)
    out_file  = File.join(tempdir, File.basename(in_file))
    files = [out_file]

    File.open(in_file) do |in_fh|
      File.open(out_file, 'w') do |out_fh|
        @announced = {}
        puts "Processing #{in_file}:"

        in_fh.each_line do |line|
          line.chomp!

          case line
          when /^Architecture: /
            line = $& + process_architectures($')
          when /^Distribution: /
            line = $& + process_distributions($')
          when /^\s [0-9a-f]{32,} \s \d+ (?:\s [a-z]+){0,2} \s/x
            file = $'
            if accept_file?(file)
              path = File.join(base_path, file)
              files << copy_file(path, tempdir)
            else
              line = nil
            end
          end

          out_fh.puts(line) unless line.nil?
        end
      end
    end

    files.compact.uniq
  end

  def process_architectures(arches)
    archlist = arches.split(/\s+/)

    if archlist.include?('source')
      raise "Source packages are not currently supported"
    end

    invalid = archlist - VALID_ARCHES
    raise "Invalid/unknown architecture(s): #{invalid.inspect}" unless invalid.empty?
    arches
  end

  def process_distributions(dists)
    distlist = dists.split(/\s+/)
    invalid  = distlist - VALID_DISTS
    raise "Invalid/unknown distribution(s): #{invalid.inspect}" unless invalid.empty?
    dists
  end

  def accept_file?(file)
    if file.end_with?('.deb')
      true
    elsif file.end_with?('.ddeb') # debug symbols
      false
    elsif file.end_with?('.dsc') || file.end_with?('.tar.gz')
      raise "Source packages are not currrently supported: #{file.inspect}"
    end
  end

  def copy_file(file, tempdir)
    target = File.join(tempdir, File.basename(file))
    return if File.exists?(target)

    puts "\t#{file}"
    FileUtils.copy(file, target)
    target
  end
end

AptUpload.new.run(*ARGV)


main.go
/*
sv-rollout is a utility to restart multiple runit services concurrently. It
supports canaries and has configurable tolerance for timeouts.
*/
package main

import (
	"flag"
	"fmt"
	"github.com/Shopify/go-dogstatsd"
	"log"
	"math/rand"
	"os"
	"os/exec"
	"path"
	"path/filepath"
	"time"
)

const (
	svdir = "/etc/service"
)

type config struct {
	CanaryRatio            float64
	CanaryTimeoutTolerance float64
	ChunkRatio             float64
	TimeoutTolerance       float64
	Timeout                int
	OnComplete             string
}

func init() {
	log.SetOutput(os.Stdout)
}

// Verbose controls whether extra information is printed as choices are made.
// It's set by the `-verbose` CLI flag.
var Verbose bool

func (c config) AssertValid(pattern string) {
	msg := ""
	if pattern == "" {
		msg = "-pattern must be provided"
	}
	if c.ChunkRatio < c.CanaryRatio {
		msg = "-chunk-ratio must be >= -canary-ratio. This is not an inherent limitation, feel free to add code to handle this case."
	}
	if msg != "" {
		fmt.Println(msg)
		flag.Usage()
		os.Exit(1)
	}
}

var (
	// Statsd is a globally-shared datadog client.
	Statsd *dogstatsd.Client
)

func configureStatsd() {
	var err error
	if Statsd != nil {
		return
	}
	Statsd, err = dogstatsd.New("127.0.0.1:8125", &dogstatsd.Context{
		Namespace: "sv-rollout.",
	})
	if err != nil {
		log.Println("Can't connect to statsd: ", err)
	}
}

func init() {
	rand.Seed(time.Now().UnixNano()) // init(), not main(), so that tests get it too.

	oldUsage := flag.Usage
	flag.Usage = func() {
		oldUsage()
		fmt.Fprintf(os.Stderr, "Examples:\n")
		fmt.Fprintln(os.Stderr, "  # Restart one service first. Restart everything else once it succeeds. No timeouts allowed, wait up to 5 minutes for restarts.")
		fmt.Fprintln(os.Stderr, "  "+os.Args[0]+" -canary-ratio 0.0001 -chunk-ratio 1 -timeout 300 -pattern 'borg-*'")
		fmt.Fprintf(os.Stderr, "%s", "  # Restart 10% of services first, allowing up to 50% of those to time out. Then, restart all other services, 30% at a time, allowing up to 70% to time out.\n")
		fmt.Fprintln(os.Stderr, "  "+os.Args[0]+" -canary-ratio 0.1 -chunk-ratio 0.3 -canary-timeout-tolerance 0.5 -timeout-tolerance 0.7 -timeout 300 -pattern 'borg-*'")
	}
}

func main() {
	configureStatsd()
	var (
		canaryRatio            = flag.Float64("canary-ratio", 0.001, "canary nodes are restarted first. If they fail, the deploy is failed. Rounded up to the nearest node, unless set to zero")
		canaryTimeoutTolerance = flag.Float64("canary-timeout-tolerance", 0, "ratio of canary nodes that are permitted to time out without causing the deploy to fail")
		chunkRatio             = flag.Float64("chunk-ratio", 0.2, "after canary nodes, ratio of remaining nodes permitted to restart concurrently")
		timeoutTolerance       = flag.Float64("timeout-tolerance", 0, "ratio of total nodes whose restarts may time out and still consider the deploy a success")
		timeout                = flag.Int("timeout", 90, "number of seconds to wait for a service to restart before considering it timed out and moving on")
		pattern                = flag.String("pattern", "", "(required) glob pattern to match /etc/service entries (e.g. \"borg-shopify-*\")")
		onComplete             = flag.String("oncomplete", "", "command to execute when the deploy finishes (regardless of success)")
		verbose                = flag.Bool("verbose", false, "print more information about what's going on")
	)

	flag.Parse()
	config := config{
		CanaryRatio:            *canaryRatio,
		CanaryTimeoutTolerance: *canaryTimeoutTolerance,
		ChunkRatio:             *chunkRatio,
		TimeoutTolerance:       *timeoutTolerance,
		Timeout:                *timeout,
		OnComplete:             *onComplete,
	}
	config.AssertValid(*pattern)

	Verbose = *verbose

	if Verbose {
		log.Printf("[debug] initializing with pattern=%s, config %#v", *pattern, config)
	}

	os.Exit(run(*pattern, config))
}

func run(servicePattern string, c config) int {
	services, err := getServices(servicePattern)
	if err != nil {
		log.Fatal(err)
	}

	defer runCompletionHandler(c.OnComplete)

	d := NewDeployment(services, c)
	if err := d.Run(); err != nil {
		return 1
	}
	return 0
}

func getServices(pattern string) (services []string, err error) {
	var fullpaths []string
	fullpaths, err = globServices(svdir + "/" + pattern)
	if err != nil {
		return nil, err
	}
	for _, p := range fullpaths {
		services = append(services, path.Base(p))
	}
	shuffle(services)
	return
}

// Fisher-Yates, Sattolo's algorithm
func shuffle(a []string) {
	for i := range a {
		j := rand.Intn(i + 1)
		a[i], a[j] = a[j], a[i]
	}
}

func runCompletionHandler(command string) {
	if command == "" {
		return
	}
	cmd := exec.Command("sh", "-c", command)
	output, err := cmd.CombinedOutput()
	if err != nil {
		log.Println("completion handler failed:", err)
		return
	}
	log.Println("completion handler:", command)
	log.Println(string(output))
}

var globServices = filepath.Glob
